### PR TITLE
ArC - log exceptions thrown by async observers

### PR DESCRIPTION
**Pull request by mkouba**
- similarly to transactional observers
- related to #23713

### CODE REVIEW

**Comment by famod** [comment]
Is there a way to opt out of that _except_ silencing the entire logger? Reason I'm asking is that we are relying on custom exception formatting (fewer stackframes) via `.exceptionally(...)`.

**Comment by mkouba** [comment]
> Is there a way to opt out of that _except_ silencing the entire logger? Reason I'm asking is that we are relying on custom exception formatting (fewer stackframes) via `.exceptionally(...)`.

Nope, but it merely logs a simple error message without the stack trace by default (the stack trace is only visible if TRACE level is enabled).

**Comment by mkouba** [comment]
> I /think/ I'm doing the same as @famod in Quarkus GitHub App to propagate the errors and I wonder if this should be optional (and maybe not enabled by default). Because typically, the error won't have the proper context needed to properly debug what's going on.

Well, it merely logs a simple message. We do the same for transactional observers (e.g. `@Observes(during = TransactionPhase.AFTER_SUCCESS`). The truth is that for transactional observers there is no way to handle the exceptions at all (unlike for `@ObservesAsync`). Still I don't think that an error message like this would annoy anyone...

**Comment by famod** [comment]
> Still I don't think that an error message like this would annoy anyone...

In a loaded environment, each "duplicate" exception report adds noise.

> The truth is that for transactional observers there is no way to handle the exceptions at all 

What about an interceptor? We actually have a custom `@CatchAll` interceptor that is bound very early (not everywhere of course, only for dedicated cases).

I'm not against this in general, I'm just asking for an opt out.

**Comment by mkouba** [comment]
> What about an interceptor? We actually have a custom `@CatchAll` interceptor that is bound very early (not everywhere of course, only for dedicated cases).

An interceptor would only work if you control the source code of all observers.

> In a loaded environment, each "duplicate" exception report adds noise.

Well, we're talking about observers that throw exceptions during notifications... which should be... exceptional?

> I'm not against this in general, I'm just asking for an opt out.

I get your point but I'm not sure it's worth a new config property. And you can always silence the logger for the `CollectingExceptionHandler` only.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java [hunk 1]
```diff
@@ -464,9 +464,11 @@ public void run() {
                 }
             } catch (Exception e) {
                 // swallow exception and log errors for every problematic OM
-                LOGGER.errorf("Failure while notifying an observer %s for event %s. Stack trace - %s",
-                        observerMethod, eventContext.getMetadata().getType().getTypeName(),
-                        e.getCause() != null ? e.getCause() : e);
+                LOG.errorf(
+                        "Failure occurred while notifying a transational %s for event of type %s \n- please enable debug logging to see the full stack trace",
+                        observerMethod, eventContext.getMetadata().getType().getTypeName());
+                LOG.debugf(e, "Failure occurred while notifying a transational %s for event of type %s",
```
Comment (manovotn) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:470]: Shouldn't this message contain the stack trace (`e.getCause()`)?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java [hunk 2]
```diff
@@ -551,19 +553,22 @@ public static Status valueOf(TransactionPhase transactionPhase) {
...
         }
 
         @Override
-        public void handle(Throwable throwable) {
+        public void handle(Throwable throwable, ObserverMethod<?> observerMethod, EventContext<?> eventContext) {
             throwables.add(throwable);
+            LOG.errorf(
+                    "Failure occurred while notifying an async %s for event of type %s \n- please enable debug logging to see the full stack trace",
+                    observerMethod, eventContext.getMetadata().getType().getTypeName());
+            LOG.debugf(throwable, "Failure occurred while notifying an async %s for event of type %s",
```
Comment (manovotn) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:570]: Same as above; shouldn't this contains the track trace?

**Comment by famod** [comment]
> > I'm not against this in general, I'm just asking for an opt out.
> 
> I get your point but I'm not sure it's worth a new config property. And you can always silence the logger for the `CollectingExceptionHandler` only.

`CollectingExceptionHandler` looks well scoped, so that would work for me.

**Comment by Ladicek** [comment]
I think this generally makes sense. When people use async events in a fire-and-forget manner, logging is probably the only general exception handling strategy we can implement.

There's one other thing we could could do, though I'm not sure how easy or hard it would be. The `Event.fireAsync()` method returns a `CompletionStage`, and we could probably detect whether someone attached a handler to it (that's obviously racy with the actual observer execution). If a handler _was_ attached, we probably don't need to log the exception, but if a handler was _not_ attached [before completion], we definitely should log the exception.

Perhaps that's too smart, though.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:470]: I don't think so - we pass the exception as an argument so it's up to the logging config...

**Comment by mkouba** [comment]
> There's one other thing we could could do, though I'm not sure how easy or hard it would be.

I think that it would be doable but I'm not sure it'd be worth the effort. Also it might be confusing if an observer is executed before a "handler is attached", i.e. in some cases we would log a warning or not depending on the timing...

**Comment by Ladicek** [comment]
Yea I mentioned that this approach would be racy. My reasoning was that the observer execution probably takes some time (otherwise why bother making it async?), while a handler is typically attached to a `CompletionStage` immediately, so it would usually work fine.

Here's another idea:

```java
public interface AsyncEventExceptionHandler {
    void handle(Throwable exception);
}
```

If the application has a bean whose set of bean types contains this interface, that bean is notified (by calling the `handle` method) of any exception thrown by an async observer. (If there's multiple beans of this type, either all of them are notified, or the one with highest priority, or something like that. Alternatively, this could also be modeled as a CDI event! :laughing:)

**Comment by mkouba** [comment]
> Yea I mentioned that this approach would be racy.

Hm, "racy" as in https://dictionary.cambridge.org/dictionary/english/racy ;-)

> My reasoning was that the observer execution probably takes some time (otherwise why bother making it async?), while a handler is typically attached to a `CompletionStage` immediately, so it would usually work fine.

I get your point. However, I don't think it's always about the "time". It might be some fast I/O operation or something that takes some time only in some cases.

> 
> Here's another idea:
> 
> ```java
> public interface AsyncEventExceptionHandler {
>     void handle(Throwable exception);
> }
> ```
> 
> If the application has a bean whose set of bean types contains this interface, that bean is notified (by calling the `handle` method) of any exception thrown by an async observer. (If there's multiple beans of this type, either all of them are notified, or the one with highest priority, or something like that. 

I like this idea!

> Alternatively, this could also be modeled as a CDI event! laughing)

I don't like this idea ;-)

**Comment by mkouba** [comment]
@Ladicek @famod What about now?

**Comment by mkouba** [comment]
Hm, I did run the formatter. @gsmet Do you have an idea what's going on?

**Review by Ladicek — APPROVED** [review]
LGTM

**Comment by famod** [comment]
@mkouba
> Hm, I did run the formatter. @gsmet Do you have an idea what's going on?

Can you try the command from this comment? https://github.com/quarkusio/quarkus/pull/24047#issuecomment-1055969577
The formatter plugin was updated recently and they changed something w.r.t. cache handling...

**Comment by mkouba** [comment]
> @mkouba
> 
> > Hm, I did run the formatter. @gsmet Do you have an idea what's going on?
> 
> Can you try the command from this comment? [#24047 (comment)](https://github.com/quarkusio/quarkus/pull/24047#issuecomment-1055969577) The formatter plugin was updated recently and they changed something w.r.t. cache handling...

Thanks. The branch was not rebased.

**Pull request merged**